rust/agents/relayer/src/merkle_tree_builder.rs

```diff
@@ -8,20 +8,63 @@ use color_eyre::eyre::Result;
 use ethers::core::types::H256;
 use std::fmt::Display;
 
-use tracing::{debug, error};
+use tracing::{debug, error, info};
+
+// Helper type to denote the fact that a Inbox that has received no checkpoint is different from an Inbox that has been checkpointed with checkpoint index 0
+#[derive(Debug, Copy, Clone)]
+pub struct OnchainCheckpointIndex {
+    index: Option<u32>,
+}
+
+impl OnchainCheckpointIndex {
+    pub fn from_checkpoint(checkpoint: &Checkpoint) -> Self {
+        Self {
+            index: if checkpoint.root.is_zero() {
+                None
+            } else {
+                Some(checkpoint.index)
+            },
+        }
+    }
+
+    pub fn from_index(index: u32) -> Self {
+        Self { index: Some(index) }
+    }
+
+    pub fn matches_signed_checkpoint(&self, other_checkpoint_index: u32) -> bool {
```

Comment:
I think this name is a little confusing because (1) it doesn't seem like there's anything specific to signed checkpoints and (2) I'm not sure if "matches" is the right word here as it's really checking if self.index >= the other checkpoint